# <center>ĐỒ ÁN CUỐI KỲ: KHOA HỌC DỮ LIỆU ỨNG DỤNG</center>

### <center>**Đề tài:** Dự đoán kết quả học tập của học sinh dựa trên việc chơi game</center>

**Nhóm:** Bumblebee - Nhóm 2

**Thành viên**

| MSSV     | Họ và tên       | Github                      |
|----------|-----------------|-----------------------------|
| 19120511 | Võ Văn Hiếu     | https://github.com/vv-hieu  |
| 19120526 | Huỳnh Đức Huy   | https://github.com/KaiKush  |
| 19120539 | Vương Thế Khang | https://github.com/vt-khang |


**Github repository:** https://github.com/vt-khang/KHDLUD-FinalProject

## Import

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set option
pd.set_option('display.max_colwidth', 999)
pd.set_option('float_format', '{:f}'.format)

## A. THU THẬP DỮ LIỆU

### Mục tiêu đề tài

Cuộc thi này nhằm mục đích xác định các cách để học sinh tham gia vào nội dung giáo dục bằng hinh thức trò chơi giáo dục, từ đó đưa ra các dự đoán hiệu suất học tập của học sinh. Trong đồ án này sử dụng trò chơi giáo dục trực tuyến [Jo Wilder](https://pbswisconsineducation.org/jowilder/play-the-game) để theo dõi quá trình học tập của học sinh từ lớp 3-5, chủ yếu về các chủ đề lịch sử. Dưới đây là một số yếu tố cốt lõi của trò chơi:
- Nhân vật: Trong trò chơi này, bạn phải tương tác với nhân vật chính cùng với các nhân vật khác nhau và ghé thăm nhiều địa điểm khác nhau trên bản đồ để tìm ra manh mối, từ đó giải đáp những bí ẩn liên quan đến lịch sử Wisconsin.
- Điều khiển: Để di chuyển qua các khung cảnh khác nhau, bạn phải sử dụng chuột và nhấp vào nơi bạn muốn di chuyển, cũng tương tự như muốn tương tác với các nhân vật.
- Notebook: Đây là nơi cung cấp thông tin và lưu những phát hiện của từng manh mối bạn có được trong trò chơi.
- Thời gian: Là thời gian bạn dành cho mỗi địa điểm khác nhau tùy thuộc vào cách bạn theo dõi câu chuyện của trò chơi.

### Giới thiệu bộ dữ liệu

Bộ dữ liệu của cuộc thi là dữ liệu về chuỗi thời gian được ghi lại từ trò chơi giáo dục trực tuyến [Jo Wilder](https://pbswisconsineducation.org/jowilder/play-the-game). Trò chơi được chia thành 4 chương, sau mỗi chương người chơi sẽ được trả lời một số câu hỏi. Có tổng cộng 18 câu hỏi liên quan trực tiếp trong trò chơi. Mục tiêu của cuộc thi là sử dụng bộ dữ liệu từ những câu hỏi trên để xác định xem người chơi có trả lời đúng câu hỏi hay không, từ đó có thể đưa ra các kết luận về hiệu suất học tập của học sinh.

Các file dữ liệu này bao gồm:
- **train.csv** - tập train
- **test.csv** - tập test
- **sample_submission.csv** - định dạng file dùng để nộp trên Kaggle
- **train_labels.csv** - câu trả lời cho 18 câu hỏi của mỗi session trong tập train

Cuộc thi: [Predict Student Performance from Game Play](https://www.kaggle.com/competitions/predict-student-performance-from-game-play) \
Nhà tổ chức cuộc thi: [Field Day Lab](https://fielddaylab.wisc.edu) \
Nhà tài trợ: [The Learning Agency Lab](https://www.the-learning-agency-lab.com) \
License: OPEN SOURCE - MIT

### Ứng dụng và ý nghĩa

- Thúc đẩy nghiên cứu về các phương pháp giáo dục, dạy học, tiếp thu kiến thức, ... từ các trò chơi giáo dục
- Hỗ trợ các nhà phát triển phần mềm trò chơi (Game Developer) để tạo ra thêm nhiều trải nghiệm học tập có hiệu quả hơn cho học sinh
- Thấy được các lợi ích rộng rãi của các nền tảng học tập dựa trên trò chơi đến với nhiều người hơn

## B. KHÁM PHÁ DỮ LIỆU

Đầu tiên, ta sẽ xem thử bộ dữ liệu trông như thế nào. Đọc dữ liệu từ file "train.csv" và lưu vào dataframe `df`.

In [2]:
df = pd.read_csv('data/train.csv')
df.head()

session_id  index  elapsed_time      event_name   name  level  page  \
0  20090312431273200      0             0  cutscene_click  basic      0   NaN   
1  20090312431273200      1          1323    person_click  basic      0   NaN   
2  20090312431273200      2           831    person_click  basic      0   NaN   
3  20090312431273200      3          1147    person_click  basic      0   NaN   
4  20090312431273200      4          1863    person_click  basic      0   NaN   

   room_coor_x  room_coor_y  screen_coor_x  screen_coor_y  hover_duration  \
0  -413.991405  -159.314686     380.000000     494.000000             NaN   
1  -413.991405  -159.314686     380.000000     494.000000             NaN   
2  -413.991405  -159.314686     380.000000     494.000000             NaN   
3  -413.991405  -159.314686     380.000000     494.000000             NaN   
4  -412.991405  -159.314686     381.000000     494.000000             NaN   

                            text    fqid                       room_fqid  \
0                      undefined   intro  tunic.historicalsociety.closet   
1  Whatcha doing over there, Jo?  gramps  tunic.historicalsociety.closet   
2         Just talking to Teddy.  gramps  tunic.historicalsociety.closet   
3     I gotta run to my meeting!  gramps  tunic.historicalsociety.closet   
4            Can I come, Gramps?  gramps  tunic.historicalsociety.closet   

                                            text_fqid  fullscreen  hq  music  \
0                tunic.historicalsociety.closet.intro           0   0      1   
1  tunic.historicalsociety.closet.gramps.intro_0_cs_0           0   0      1   
2  tunic.historicalsociety.closet.gramps.intro_0_cs_0           0   0      1   
3  tunic.historicalsociety.closet.gramps.intro_0_cs_0           0   0      1   
4  tunic.historicalsociety.closet.gramps.intro_0_cs_0           0   0      1   

  level_group  
0         0-4  
1         0-4  
2         0-4  
3         0-4  
4         0-4

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [3]:
df.shape

(26296946, 20)

Vậy dữ liệu có kích thước **26,296,946 dòng x 20 cột**.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Quan sát sơ bộ dữ liệu ta thấy mỗi dòng chứa thông tin về một sự kiện trong một session của trò chơi, có vẻ như không có vấn đề các dòng có ý nghĩa khác nhau.

### Dữ liệu có các dòng bị lặp không?

Kiểm tra xem dữ liệu có các dòng bị lặp không và lưu kết quả vào biến `have_duplicated_row`. Biến này sẽ có giá trị True nếu dữ liệu có các dòng bị lặp và có giá trị False nếu ngược lại.

In [4]:
have_duplicated_row = any(df.duplicated())
have_duplicated_row

False

Như vậy, không có dòng nào bị lặp lại.

### Mỗi cột có ý nghĩa gì?

Mỗi cột có ý nghĩa như sau:
-	**session_id** – id của session sự kiện xảy ra
-	**index** – index của sự kiện cho session, dùng để phân biệt trong cùng một session
-	**elapsed_time** – khoảng thời gian đã trôi qua (tính bằng ms) giữa thời điểm bắt đầu session cho đến khi sự kiện được ghi lại
-	**event_name** – tên của loại sự kiện
-	**name** – tên sự kiện (ví dụ: xác định xem sự kiện `notebook_click` là đang mở hay đóng cuốn sổ ghi chép)
-	**level** – sự kiện xảy ra ở level nào của trò chơi (từ 0 đến 22)
-	**page** – số trang của sự kiện (chỉ dành cho các sự kiện có liên quan đến cuốn sổ ghi chép)
-	**room_coor_x** – tọa độ x của lần click vào room trong trò chơi (chỉ dành cho các sự kiện click)
-	**room_coor_y** – tọa độ y của lần click vào room trong trò chơi (chỉ dành cho các sự kiện click)
-	**screen_coor_x** – tọa độ x của lần click liên quan đến màn hình của người chơi (chỉ dành cho các sự kiện click)
-	**screen_coor_y** – tọa độ y của lần click liên quan đến màn hình của người chơi (chỉ dành cho các sự kiện click)
-	**hover_duration** – thời lượng hover  (tính bằng ms) xảy ra trong bao lâu (chỉ dành cho các sự kiện hover)
-	**text** – văn bản mà người chơi nhìn thấy trong sự kiện này
-	**fqid** – id đầy đủ của sự kiện
-	**room_fqid** – id đầy đủ của room mà sự kiện xảy ra
-	**text_fqid** – id đầy đủ của văn bản
-	**fullscreen** – trò chơi có được mở chế độ toàn màn hình không
-	**hq** – trò chơi có bật chất lượng cao hay không
-	**music** – có bật hay tắt nhạc trong trò chơi không
-	**level_group** – nhóm các level và nhóm các câu hỏi (0-4, 5-12, 13-22)

### Mỗi cột hiện đang có kiểu dữ liệu gì?

Xem thử kiểu dữ liệu của các cột dữ liệu.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26296946 entries, 0 to 26296945
Data columns (total 20 columns):
 #   Column          Dtype  
---  ------          -----  
 0   session_id      int64  
 1   index           int64  
 2   elapsed_time    int64  
 3   event_name      object 
 4   name            object 
 5   level           int64  
 6   page            float64
 7   room_coor_x     float64
 8   room_coor_y     float64
 9   screen_coor_x   float64
 10  screen_coor_y   float64
 11  hover_duration  float64
 12  text            object 
 13  fqid            object 
 14  room_fqid       object 
 15  text_fqid       object 
 16  fullscreen      int64  
 17  hq              int64  
 18  music           int64  
 19  level_group     object 
dtypes: float64(6), int64(7), object(7)
memory usage: 3.9+ GB


Như vậy có thể thấy:
- Các cột "session_id", "index", "level", "fullscreen", "hq", "music" đang có kiểu dữ liệu `int`. Ta cần đưa các cột này về dạng `category`.
- Các cột "event_name", "name", "fqid", "room_fqid", "text_fqid", "level_group" đang có kiểu dữ liệu `object`. Ta cần đưa các cột này về dạng `category`.
- Cột "page" đang có kiểu dữ liệu `float`. Ta cần đưa cột này về dạng `category`.

---

**Tiền xử lý dữ liệu**: Thay đổi kiểu dữ liệu

Thay đổi kiểu dữ liệu của các cột "session_id", "index", "level", "fullscreen", "hq", "music" sang kiểu dữ liệu dạng `category`.

In [6]:
df[['session_id', 'index', 'level', 'fullscreen', 'hq', 'music']] = df[['session_id', 'index', 'level', 'fullscreen', 'hq', 'music']].astype('category')

Thay đổi kiểu dữ liệu của các cột "event_name", "name", "fqid", "room_fqid", "text_fqid", "level_group" sang kiểu dữ liệu dạng `category`.

In [7]:
df[['event_name', 'name', 'fqid', 'room_fqid', 'text_fqid', 'level_group']] = df[['event_name', 'name', 'fqid', 'room_fqid', 'text_fqid', 'level_group']].astype('category')

Thay đổi kiểu dữ liệu của các cột "page" sang kiểu dữ liệu dạng `category`.

In [8]:
df['page'] = df['page'].astype('Int64').astype('category')

Xem thử kiểu dữ liệu của các cột dữ liệu sau khi tiền xử lý.

In [9]:
df.dtypes

session_id        category
index             category
elapsed_time         int64
event_name        category
name              category
level             category
page              category
room_coor_x        float64
room_coor_y        float64
screen_coor_x      float64
screen_coor_y      float64
hover_duration     float64
text                object
fqid              category
room_fqid         category
text_fqid         category
fullscreen        category
hq                category
music             category
level_group       category
dtype: object

---

Có vẻ các cột đều có kiểu dữ liệu phù hợp. Nếu trong quá trình phân tích dữ liệu cần phải thay đổi kiểu dữ liệu của các cột thì ta sẽ quay lại tiền xử lý ở đây, tạm thời ta chấp nhận kiểu dữ liệu hiện tại của các cột.

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Với mỗi cột numerical, ta sẽ tính số lượng giá trị thiếu, tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị xuất hiện (count), giá trị nhỏ nhất (min), giá trị trung bình (mean), giá trị trung vị (median), giá trị lớn nhất (max), các khoảng tứ phân vị (Q1, Q2, Q3) và độ lệch chuẩn (std). Sau đó ta sẽ lưu các giá trị này vào dataframe `numerical_col_df` thành 11 dòng là "num_missing_val", "ratio_missing_val", "count", "min", "mean", "median", "max", "Q1", "Q2", "Q3", "std".

In [10]:
def num_missing_val(s):
  return s.isna().sum()
def ratio_missing_val(s):
  return s.isna().mean() * 100
def count(s):
  return s.count();
def Q1(s):
  return s.quantile(0.25)
def Q2(s):
  return s.quantile(0.5)
def Q3(s):
  return s.quantile(0.75)

numerical_cols = df.select_dtypes(include = 'number').columns
numerical_col_df = df[numerical_cols].agg([num_missing_val, ratio_missing_val, count, min, np.mean, np.median, max, Q1, Q2, Q3, np.std]).T
numerical_col_df[['num_missing_val', 'count']] = numerical_col_df[['num_missing_val', 'count']].astype(int)
numerical_col_df

num_missing_val  ratio_missing_val     count          min  \
elapsed_time                  0           0.000000  26296946     0.000000   
room_coor_x             2073272           7.884079  24223674 -1992.354569   
room_coor_y             2073272           7.884079  24223674  -918.162349   
screen_coor_x           2073272           7.884079  24223674     0.000000   
screen_coor_y           2073272           7.884079  24223674     0.000000   
hover_duration         24294702          92.386021   2002244     0.000000   

                         mean         median               max            Q1  \
elapsed_time   4016693.894748 1014801.000000 1988606704.000000 439700.000000   
room_coor_x        -54.905793     -11.162797       1261.773745   -352.839501   
room_coor_y       -116.211852     -97.605339        543.616424   -212.718256   
screen_coor_x      458.409812     447.000000       1919.000000    269.000000   
screen_coor_y      385.586034     397.000000       1440.000000    304.000000   
hover_duration    3471.003377     419.000000  245163690.000000     99.000000   

                           Q2             Q3             std  
elapsed_time   1014801.000000 1747040.000000 31270320.364611  
room_coor_x        -11.162797     296.651499      520.292583  
room_coor_y        -97.605339      22.687860      218.470954  
screen_coor_x      447.000000     663.000000      247.938892  
screen_coor_y      397.000000     471.000000      129.816557  
hover_duration     419.000000    1266.000000   460202.415493

Quan sát bảng thống kê mô tả của các cột có kiểu dữ liệu dạng số (numerical), ta thấy giá trị lớn nhất của 2 cột "screen_coor_x" và "screen_coor_y" là **1919** và **1440**. Điều này có nghĩa là kích thước màn hình trò chơi chỉ đạt tối đa **1919x1440**.

Trong khi đó, ta thấy giá trị nhỏ nhất và giá trị lớn nhất của 2 cột "room_coor_x" và "room_coor_y" có mang giá trị âm và dương.

Ngoài ra, ta thấy có xuất hiện giá trị bị thiếu ở các cột "room_coor_x", "room_coor_y", "screen_coor_x", "screen_coor_y", "hover_duration", trong đó:
- Cột "hover_duration" có tỉ lệ giá trị bị thiếu nhiều nhất **(~92.39%)** so với các cột còn lại.
- Các cột "room_coor_x", "room_coor_y", "screen_coor_x", "screen_coor_y" đều có tỉ lệ giá trị bị thiếu là **7.88%**
  
Ta sẽ xem thử các dòng có chứa dữ liệu bị thiếu ở các cột này.

In [11]:
df[df['room_coor_x'].isna() | df['room_coor_y'].isna() | df['screen_coor_x'].isna() | df['screen_coor_y'].isna() | df['hover_duration'].isna()]

session_id index  elapsed_time      event_name       name  \
0         20090312431273200     0             0  cutscene_click      basic   
1         20090312431273200     1          1323    person_click      basic   
2         20090312431273200     2           831    person_click      basic   
3         20090312431273200     3          1147    person_click      basic   
4         20090312431273200     4          1863    person_click      basic   
...                     ...   ...           ...             ...        ...   
26296941  22100221145014656  1600       5483231  navigate_click  undefined   
26296942  22100221145014656  1601       5485166  navigate_click  undefined   
26296943  22100221145014656  1602       5485917  navigate_click  undefined   
26296944  22100221145014656  1603       5486753  navigate_click  undefined   
26296945  22100221145014656  1604       5487952      checkpoint      basic   

         level page  room_coor_x  room_coor_y  screen_coor_x  screen_coor_y  \
0            0  NaN  -413.991405  -159.314686     380.000000     494.000000   
1            0  NaN  -413.991405  -159.314686     380.000000     494.000000   
2            0  NaN  -413.991405  -159.314686     380.000000     494.000000   
3            0  NaN  -413.991405  -159.314686     380.000000     494.000000   
4            0  NaN  -412.991405  -159.314686     381.000000     494.000000   
...        ...  ...          ...          ...            ...            ...   
26296941    22  NaN   343.887291    36.701026     483.000000     273.000000   
26296942    22  NaN   332.696070   141.493178     545.000000     221.000000   
26296943    22  NaN   369.912859   140.569205     611.000000     217.000000   
26296944    22  NaN   252.299653   123.805889     526.000000     232.000000   
26296945    22  NaN          NaN          NaN            NaN            NaN   

          hover_duration                           text            fqid  \
0                    NaN                      undefined           intro   
1                    NaN  Whatcha doing over there, Jo?          gramps   
2                    NaN         Just talking to Teddy.          gramps   
3                    NaN     I gotta run to my meeting!          gramps   
4                    NaN            Can I come, Gramps?          gramps   
...                  ...                            ...             ...   
26296941             NaN                            NaN             NaN   
26296942             NaN                            NaN  chap4_finale_c   
26296943             NaN                            NaN             NaN   
26296944             NaN                            NaN  chap4_finale_c   
26296945             NaN                            NaN  chap4_finale_c   

                               room_fqid  \
0         tunic.historicalsociety.closet   
1         tunic.historicalsociety.closet   
2         tunic.historicalsociety.closet   
3         tunic.historicalsociety.closet   
4         tunic.historicalsociety.closet   
...                                  ...   
26296941            tunic.capitol_2.hall   
26296942            tunic.capitol_2.hall   
26296943            tunic.capitol_2.hall   
26296944            tunic.capitol_2.hall   
26296945            tunic.capitol_2.hall   

                                                   text_fqid fullscreen hq  \
0                       tunic.historicalsociety.closet.intro          0  0   
1         tunic.historicalsociety.closet.gramps.intro_0_cs_0          0  0   
2         tunic.historicalsociety.closet.gramps.intro_0_cs_0          0  0   
3         tunic.historicalsociety.closet.gramps.intro_0_cs_0          0  0   
4         tunic.historicalsociety.closet.gramps.intro_0_cs_0          0  0   
...                                                      ...        ... ..   
26296941                                                 NaN          0  0   
26296942                                                 NaN          0  0   

Quan sát có thể thấy các cột "room_coor_x", "room_coor_y", "screen_coor_x", "screen_coor_y" thường bị thiếu chung với nhau, trong khi đó cột "hover_duration" chỉ thiếu khi sự kiện đó không phải là sự kiện hover. Để xác nhận điều này, ta sẽ xem thử giá trị bị thiếu ở từng loại sự kiện dựa vào cột "event_name".

In [12]:
df.groupby(['event_name']).agg(ratio_missing_val)[numerical_cols]

elapsed_time  room_coor_x  room_coor_y  screen_coor_x  \
event_name                                                                  
checkpoint              0.000000   100.000000   100.000000     100.000000   
cutscene_click          0.000000     0.000000     0.000000       0.000000   
map_click               0.000000     0.000000     0.000000       0.000000   
map_hover               0.000000   100.000000   100.000000     100.000000   
navigate_click          0.000000     0.000000     0.000000       0.000000   
notebook_click          0.000000     0.000000     0.000000       0.000000   
notification_click      0.000000     0.000000     0.000000       0.000000   
object_click            0.000000     0.000000     0.000000       0.000000   
object_hover            0.000000   100.000000   100.000000     100.000000   
observation_click       0.000000     0.000000     0.000000       0.000000   
person_click            0.000000     0.000000     0.000000       0.000000   

                    screen_coor_y  hover_duration  
event_name                                         
checkpoint             100.000000      100.000000  
cutscene_click           0.000000      100.000000  
map_click                0.000000      100.000000  
map_hover              100.000000        0.000000  
navigate_click           0.000000      100.000000  
notebook_click           0.000000      100.000000  
notification_click       0.000000      100.000000  
object_click             0.000000      100.000000  
object_hover           100.000000        0.000000  
observation_click        0.000000      100.000000  
person_click             0.000000      100.000000

Như vậy có thể thấy:
- Các sự kiện liên quan tới hover (như "map_hover", "object_hover") thì cột "hover_duration" không bao giờ bị thiếu giá trị.
- Các sự kiện liên quan tới hover (như "map_hover", "object_hover") và "checkpoint" thì các cột "room_coor_x", "room_coor_y", "screen_coor_x", "screen_coor_y" luôn luôn bị thiếu giá trị.

Ta sẽ xử lý điều này ở bước tiền xử lý giá trị bị thiếu.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?


Với mỗi cột categorical, ta sẽ tính số lượng giá trị thiếu, tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị xuất hiện (count), giá trị xuất hiện nhiều nhất (mode), số lượng giá trị xuất hiện nhiều nhất (không xét giá trị NaN), số lượng các giá trị khác nhau (không xét giá trị NaN) và danh sách các giá trị khác nhau (không xét giá trị NaN). Sau đó ta sẽ lưu các giá trị này vào dataframe `categical_col_df` thành 6 dòng là "num_missing_val", "ratio_missing_val", "count", "top", "freq", "nunique", "unique".

In [13]:
def nunique(s):
  return s.nunique()
def unique(s):
  return s.dropna().astype('object').unique()
def top(s):
  return s.mode()[0]
def freq(s):
  return (s == s.mode()[0]).value_counts()[True]

categorical_cols = df.select_dtypes(exclude = 'number').columns
categorical_col_df = df[categorical_cols].agg([num_missing_val, ratio_missing_val, count, top, freq, nunique, unique]).T
categorical_col_df[['num_missing_val', 'count']] = categorical_col_df[['num_missing_val', 'count']].astype(int)
categorical_col_df

num_missing_val ratio_missing_val     count  \
session_id                 0          0.000000  26296946   
index                      0          0.000000  26296946   
event_name                 0          0.000000  26296946   
name                       0          0.000000  26296946   
level                      0          0.000000  26296946   
page                25732402         97.853196    564544   
text                16679807         63.428685   9617139   
fqid                 8274415         31.465308  18022531   
room_fqid                  0          0.000000  26296946   
text_fqid           16679702         63.428286   9617244   
fullscreen                 0          0.000000  26296946   
hq                         0          0.000000  26296946   
music                      0          0.000000  26296946   
level_group                0          0.000000  26296946   

                                                    top      freq nunique  \
session_id                            20110310463989760     19032   23562   
index                                                76     23549   20348   
event_name                               navigate_click  11326433      11   
name                                          undefined  12705785       6   
level                                                18   3499163      23   
page                                                  5    101171       7   
text                                          undefined    355969     597   
fqid                                             worker   1878958     128   
room_fqid                 tunic.historicalsociety.entry   3622003      19   
text_fqid    tunic.historicalsociety.cage.confrontation    666557     126   
fullscreen                                            0  22694166       2   
hq                                                    0  23135929       2   
music                                                 1  24439799       2   
level_group                                       13-22  13471703       3   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             unique  
session_id   [20090312431273200, 20090312433251036, 20090312455206810, 20090313091715820, 20090313571836404, 20090314035813970, 20090314121766812, 20090314221187252, 20090314363702160, 20090314441803444, 20090315081004164, 20090315085850788, 20090315101457836, 20090315170769824, 20090315270479452, 20090316152177500, 20090316190523732, 20090317080721164, 20090317111400710, 20090317414400068, 20090318021527530, 20090318322359292, 20090318541618010, 20090319012741024, 20090320140443148, 20100000291486630, 20100000404860470, 20100006432365430, 20100007011166932, 20100007445515820, 20100008103581056, 20100008262217530, 20100008572266050, 20100009003806170, 20100009303098716, 20100009314461780, 20100009385330468, 20100009560590796, 20100010540710908, 20100010554738616, 20100011003256550, 20100011033647636, 20100011050028640, 20100011274731204, 20100011290663880, 20100011394713344, 20100011571871700, 20100012014517956, 2010001203506972